pipeline/psth.py

```diff
@@ -19,129 +22,229 @@
+        return cls.get_func({'trial_condition_desc': trial_condition_desc})()
+
+    @classmethod
+    def get_func(cls, key):
+        self = cls()
+
+        func, args = (self & key).fetch1(
+            'trial_condition_func', 'trial_condition_arg')
+
+        return partial(dict(getmembers(cls))[func], **args)
+
+    @classmethod
+    def _get_trials_no_stim(cls, task=None, task_protocol=None, outcome=None,
+                            early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_no_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
+                 & {'trial_instruction': trial_instruction}
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) - experiment.PhotostimEvent)
+
+    @classmethod
+    def _get_trials_stim(cls, task=None, task_protocol=None, outcome=None,
+                         early_lick=None, trial_instruction=None):
+
+        log.debug('_get_trials_stim', locals())
+
+        return ((experiment.BehaviorTrial
+                 & {'task': task}
+                 & {'trial_instruction': trial_instruction}
+                 & {'early_lick': early_lick}
+                 & {'outcome': outcome}) & experiment.PhotostimEvent)
+
+
+@schema
+class UnitPsth(dj.Computed):
+    definition = """
+    -> TrialCondition
+    -> ephys.Unit
+    ---
+    unit_psth=NULL:                             longblob
+    """
     psth_params = {'xmin': -3, 'xmax': 3, 'binsize': 0.04}
 
-    @staticmethod
-    def compute_unit_trial_psth(unit_key, trial_keys):
+    key_source = (ephys.Unit * TrialCondition
+                  & [{'trial_condition_desc': desc}
+                     for desc in ['good_noearlylick_left_hit',
+                                  'good_noearlylick_right_hit',
+                                  'good_noearlylick_left_miss',
+                                  'good_noearlylick_right_miss']])
 
-        q = (ephys.TrialSpikes() & unit_key & trial_keys)
+    def make(self, key):
+        log.info('UnitPsth.make(): key: {}'.format(key))
+
+        # build unit key,
+        unit = {k: v for k, v in key.items() if k in ephys.Unit.primary_key}
+
+        # expand TrialCondition to trials,
+        trials = TrialCondition.get_trials(
+            (TrialCondition & key).fetch1('trial_condition_desc'))
+
+        # fetch related spike times
+        q = (ephys.TrialSpikes & unit & trials)
         spikes = q.fetch('spike_times')
-        return UnitPsth.compute_psth(spikes)
 
-    @staticmethod
-    def compute_psth(session_unit_spikes):
-        spikes = np.concatenate(session_unit_spikes)
+        if len(spikes) == 0:
+            log.warning('no spikes found for key {} - null psth'.format(key))
+            self.insert1(key)
+            return
 
-        xmin, xmax, bins = UnitPsth.psth_params.values()
+        spikes = np.concatenate(spikes)
+
+        xmin, xmax, bins = self.psth_params.values()
```

Comment:
Why is not using the unit psth function?

Reply:
was thinking this function is not needed with the table present (see NotImplementedErrors in not-yet-ported-or-deleted functions below). Will incorporate or not through the remaining plotting. I suppose no harm in having a function though..

Reply:
implemented via function; whynot.